draft-ietf-jsonpath-base.md

```diff
@@ -1594,7 +1594,12 @@ $[?count(@.*.author) >= 5]
 Its only argument is a nodelist.
 The result is a value, an unsigned integer, that gives the number of
 nodes in the nodelist.
-Note: there is no deduplication of the nodelist.
+
+Notes:
+
+* There is no deduplication of the nodelist.
+* There is one node in a nodelist consisting of a single node, even if that node is an array or an object
+with multiple children. For example, `count(@)` always equals 1, as does `count($)`.
```

Comment:
I think we need to focus more on that `@` itself returns a single-node nodelist.  `count(@)==1` is then a mere consequence of this, and it becomes apparent why this note is warranted.

```suggestion
* `@` always returns a nodelist of length 1.  Therefore `count(@)` always equals 1.  To obtain the number of children of the current node, the subquery requires additional segments, e.g. `count(@.*)`.
```

This removes the redundant language that @cabo was referring to.

Side note: I don't think it's necessary to include `count($)` here, but leave it if you think it helps.

Reply:
It would seem strange to start explaining the semantics of `@` here. If that needs doing, it would be better to do it in the definition of `@`.

I would prefer to keep the text here general. We could talk about `count()` of a singular query and give `@` as an example. But of course, `@.foo`, `@[23]`, etc. are equally valid examples and I'm not sure what's so special about `@`.

Reply:
I think the distinction is in the difference between how `length(@)` and `count(@)` operate, and the note needs to focus on that.  It's less about `@` itself, and more about how it's processed by each function.

For `length(@)`, the singular node's value is extracted from the nodelist and the "length" of that value is returned.  If the value is an object or array, then this is the number of children of that object or array.

But for `count(@)` that value extraction doesn't occur, even though you still have a singular nodelist.  The "count" then is the number of nodes in the nodelist, which is just one.

This distinction is what originally caught me.  Once I realized the distinction, I knew how to fix the queries.

It may seem obvious to you because you've thought through it, likely many times, but I assure you that it's not going to be immediately obvious to most people.  The confusion is compounded by the fact that many people will regard the nodelist as an array (which I know isn't right, but that won't stop people from thinking that way).

Reply:
The behaviours of `count()` and `length()` are clearly defined by the existing text, but the above difficulty could arise when readers:

(a) overlook the fact that `count()` takes a parameter of type `NodesType`, and/or
(b) forget (or fail to understand) what a nodelist is, and/or
(c) forget which function is which.

There is also the distinct possibility that people will use these functions without reading the spec at all, e.g. by cutting and pasting from an example query found on the internet.

There is nothing special about the singular query `@` and any singular query (such as `@.foo` or `$[2]`) is equally likely to cause exactly the same confusion.

One possible improvement would be to rename `count()` to something like `nodes()`, `num_nodes()`, or similar. This emphases the nodelist aspect and is perhaps more easily distinguished from `length()`.

WDYT @gregsdennis and @cabo?

Reply:
I don't think renaming `count()` would be an improvement.

I don't see a need for action here at all, but a single example might still help.  
`count(@)` would be one such example (although practically useless).

Reply:
Ok, thanks, I'm going to delete the `count($)` example and merge.

Reply:
@gregsdennis I'd like your approval before merging. I think we've nailed down `count(@)` which was your initial issue.